File: mac/Sources/LanClipCore/RelPath.swift
import Foundation

public enum RelPath {
    static let maxComponent = 150
    /// Находка I8 финального ревью: было 400, что не влезает в легаси
    /// `MAX_PATH` (260 символов, включая диск и завершающий `NUL`) вместе со
    /// стейджинг-корнем на Windows (`%LOCALAPPDATA%\lanclip\incoming\<метка>\`,
    /// на реальной машине ~66 символов, плюс запас на более длинное имя
    /// пользователя). `rel` длиной 195–400 нормализацию проходил, но на приёме
    /// падал необёрнутым `PathTooLongException`. Выбор 150 (не отдельно
    /// придуманное число, а тот же предел, что и `maxComponent`) с запасом в
    /// ~100 символов на корень партии укладывается в 260 с большим запасом на
    /// обеих платформах. Предел обязан совпадать на Mac и Windows побитово —
    /// это часть протокола, а не деталь реализации одной стороны.
    static let maxTotal = 150

    private static let illegal: Set<Character> = ["<", ">", ":", "\"", "|", "?", "*", "\0"]

    private static let reserved: Set<String> = {
        var names: Set<String> = ["con", "prn", "aux", "nul"]
        for n in 1...9 {
            names.insert("com\(n)")
            names.insert("lpt\(n)")
        }
        return names
    }()

    public static func normalize(_ raw: String) -> String? {
        let unified = raw.replacingOccurrences(of: "\\", with: "/")
        var components: [String] = []

        for piece in unified.split(separator: "/", omittingEmptySubsequences: true) {
            let component = String(piece)
            if component == "." { continue }
            if component == ".." { return nil }
            guard let safe = sanitize(component) else { return nil }
            components.append(safe)
        }

        guard !components.isEmpty else { return nil }
        let joined = components.joined(separator: "/")
        // Байты UTF-8, не графемы: та же единица, что и на Windows-стороне
        // (см. RelPath.cs) — иначе платформы расходятся на одном и том же
        // входе (эмодзи, иероглифы из дополнительных плоскостей Unicode).
        guard joined.utf8.count <= maxTotal else { return nil }
        return joined
    }

    private static func sanitize(_ component: String) -> String? {
        var cleaned = String(component.map { char in
            if illegal.contains(char) { return "_" }
            if let scalar = char.unicodeScalars.first,
               char.unicodeScalars.count == 1,
               scalar.properties.generalCategory == .control {
                return "_"
            }
            return char
        })

        while let last = cleaned.last, last == "." || last == " " {
            cleaned.removeLast()
        }

        guard !cleaned.isEmpty else { return nil }

        // Мелкая находка финального ревью: `split(separator: ".")` по умолчанию
        // ОТБРАСЫВАЕТ пустые подпоследовательности, поэтому для `".con"` (ведущая
        // точка) `base` получался равным "con" — реальный Windows так реестр
        // имён не проверяет: там "имя до первой точки" для `.con` — ПУСТАЯ строка
        // (точка стоит на нулевой позиции), а не "con". Из-за этого `.con`
        // ошибочно распознавался как зарезервированное имя и получал суффикс
        // (`.con_`), а Windows-сторонняя реализация (индекс первой точки, без
        // отбрасывания пустых кусков) корня не трогала вовсе — один и тот же файл
        // приезжал на две машины под разными именами. Взят явный индекс первой
        // точки — так же, как на Windows-стороне (`RelPath.cs`), — вместо `split`.
        let base: String
        if let dotIndex = cleaned.firstIndex(of: ".") {
            base = String(cleaned[cleaned.startIndex..<dotIndex])
        } else {
            base = cleaned
        }
        if reserved.contains(base.lowercased()) {
            cleaned += "_"
        }

        // Байты UTF-8, не графемы (см. комментарий в normalize(_:)).
        guard cleaned.utf8.count <= maxComponent else { return nil }
        return cleaned
    }
}

File: win/src/RelPath.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace LanClip
{
    static class RelPath
    {
        const int MaxComponent = 150;
        const int MaxTotal = 400;

        static readonly HashSet<char> Illegal = BuildIllegal();
        static readonly HashSet<string> Reserved = BuildReserved();

        static HashSet<char> BuildIllegal()
        {
            HashSet<char> chars = new HashSet<char>();
            chars.Add('<');
            chars.Add('>');
            chars.Add(':');
            chars.Add('"');
            chars.Add('|');
            chars.Add('?');
            chars.Add('*');
            chars.Add('\0');
            return chars;
        }

        static HashSet<string> BuildReserved()
        {
            HashSet<string> names = new HashSet<string>();
            names.Add("con");
            names.Add("prn");
            names.Add("aux");
            names.Add("nul");
            for (int n = 1; n <= 9; n++)
            {
                names.Add("com" + n);
                names.Add("lpt" + n);
            }
            return names;
        }

        // Семь правил в строго этом порядке (см. mac/Sources/LanClipCore/RelPath.swift —
        // эталон, с которым эта реализация обязана совпадать бит в бит):
        // 1. \ -> /
        // 2. разбить и выбросить пустые компоненты и "."
        // 3. любой компонент ".." -> отказ
        // 4. санитизация запрещённых символов и управляющих символов в "_"
        // 5. обрезка хвостовых точек и пробелов
        // 6. суффикс "_" к зарезервированным именам — ДО проверки длины компонента
        // 7. отказ при компоненте длиннее MaxComponent или итоговом пути длиннее MaxTotal
        public static string Normalize(string raw)
        {
            string unified = raw.Replace("\\", "/");
            string[] pieces = unified.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
            List<string> components = new List<string>();

            foreach (string piece in pieces)
            {
                if (piece == ".")
                {
                    continue;
                }
                if (piece == "..")
                {
                    return null;
                }
                string safe = Sanitize(piece);
                if (safe == null)
                {
                    return null;
                }
                components.Add(safe);
            }

            if (components.Count == 0)
            {
                return null;
            }

            string joined = string.Join("/", components.ToArray());
            if (joined.Length > MaxTotal)
            {
                return null;
            }
            return joined;
        }

        static string Sanitize(string component)
        {
            StringBuilder builder = new StringBuilder(component.Length);
            foreach (char c in component)
            {
                if (Illegal.Contains(c) || CharUnicodeInfo.GetUnicodeCategory(c) == UnicodeCategory.Control)
                {
                    builder.Append('_');
                }
                else
                {
                    builder.Append(c);
                }
            }

            string cleaned = builder.ToString();
            while (cleaned.Length > 0 && (cleaned[cleaned.Length - 1] == '.' || cleaned[cleaned.Length - 1] == ' '))
            {
                cleaned = cleaned.Substring(0, cleaned.Length - 1);
            }

            if (cleaned.Length == 0)
            {
                return null;
            }

            int dotIndex = cleaned.IndexOf('.');
            string baseName = dotIndex >= 0 ? cleaned.Substring(0, dotIndex) : cleaned;
            if (Reserved.Contains(baseName.ToLowerInvariant()))
            {
                cleaned = cleaned + "_";
            }

            if (cleaned.Length > MaxComponent)
            {
                return null;
            }
            return cleaned;
        }
    }
}
